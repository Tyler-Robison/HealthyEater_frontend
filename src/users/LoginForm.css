.LoginForm {
    background-image: url('../Images/homepage_img.jpg');
    background-repeat: no-repeat;
    background-size: cover;
    height: 94.3vh;
    display: flex;
    flex-direction: column;
    align-items: center;
}

.LoginForm-div { 
    background-color: white;
    width: 24vw;
    border-radius: 1rem;
    box-shadow: 2px 2px 8px rgba(0, 0, 0, 1);
}

.LoginForm-div input{ 
    margin-bottom: 1rem;
}



.LoginForm-btn {
    margin-bottom: 1rem;
}