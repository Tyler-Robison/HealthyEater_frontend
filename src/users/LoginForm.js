import React, { useContext, useEffect } from "react";
import UserAPI from "../APIs/userAPI";
import { useNavigate } from 'react-router-dom';
import { useFormik } from "formik";
import loginValidate from './user_validators/loginValidate'
import GlobalContext from "../context/GlobalContext";
import useTimedMessage from "../customHooks/useTimedMessage";
import './LoginForm.css'

/** LoginForm allows existing users to login
 * 
 * upon succesful login currentUser will be created, allowing all components to access user info*/
const LoginForm = () => {
    const validate = loginValidate
    const navigate = useNavigate();
    const { currentUser, login } = useContext(GlobalContext)
    const [failureMsg, setFailureMsg] = useTimedMessage()

    useEffect(() => {
        if (currentUser) navigate('/')
    })

    const formik = useFormik({
        initialValues: {
            username: '',
            password: ''
        },
        validate,
        onSubmit: values => loginUser(values),
    })

    // login sets token, which triggers useEffect in App.js
    const loginUser = async (values) => {
        try {
            const res = await UserAPI.login(values)
            login(res.token);
            navigate('/')

        } catch (err) {
            formik.resetForm();
            setFailureMsg(true)
        }
    }

    return (
        <div className="LoginForm">
            <div className="row mt-3">
                <div className="LoginForm-div col-12">
                    <h1>Enter Username/Password</h1>
                    {failureMsg && <p className="error-msg">Invalid Username/Password</p>}

                    <form onSubmit={formik.handleSubmit}>
                        <div>
                            <div><label htmlFor="username">Username</label></div>
                            <input
                                id="username"
                                name="username"
                                type='text'
                                onChange={formik.handleChange}
                                onBlur={formik.handleBlur}
                                value={formik.values.username}
                                placeholder="Username"
                            />
                            {formik.touched.username && formik.errors.username && (
                                <div className="error-msg">{formik.errors.username}</div>
                            )}
                        </div>
                        <div>
                            <div><label htmlFor="password">Password</label></div>
                            <input
                                id="password"
                                name="password"
                                type='password'
                                onChange={formik.handleChange}
                                onBlur={formik.handleBlur}
                                value={formik.values.password}
                                placeholder="Password"
                            />
                            {formik.touched.password && formik.errors.password && (
                                <div className="error-msg">{formik.errors.password}</div>
                            )}
                        </div>
                        <button className="general-btn LoginForm-btn" data-testid="login-btn" type="submit">Login</button>
                    </form>
                </div>
            </div>
        </div>
    )
}

export default LoginForm;