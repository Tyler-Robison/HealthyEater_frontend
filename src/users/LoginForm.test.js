import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import LoginForm from './LoginForm';
import { MemoryRouter } from "react-router-dom";
import ContextProvider from '../testContext'

// ability to login is tested in app.test.js
it("renders without crashing", function () {
    <MemoryRouter>
        <ContextProvider userData={null}>
            <LoginForm />
        </ContextProvider>
    </MemoryRouter>
});

it("matches snapshot", function () {

    const container = render(
        <MemoryRouter>
            <ContextProvider userData={null}>
                <LoginForm />
            </ContextProvider>
        </MemoryRouter>
    );
    expect(container.asFragment()).toMatchSnapshot();
});

it("displays correct error msg on missing password", async function () {

    render(
        <MemoryRouter>
            <ContextProvider userData={null}>
                <LoginForm />
            </ContextProvider>
        </MemoryRouter>
    );

    const loginBtn = screen.getByText('Login');
    const usernameInput = screen.getByLabelText('Username');

    fireEvent.change(usernameInput, { target: { value: 'Tyler' } });
    fireEvent.click(loginBtn);

    // login failure, correct error msg displayed
    await waitFor(() => {
        expect(screen.getByText('Password Required')).toBeInTheDocument();
    });

});

it("displays correct error msg on missing username", async function () {

    render(
        <MemoryRouter>
            <ContextProvider userData={null}>
                <LoginForm />
            </ContextProvider>
        </MemoryRouter>
    );

    const loginBtn = screen.getByText('Login');
    const passwordInput = screen.getByLabelText('Password');

    fireEvent.change(passwordInput, { target: { value: 'password' } });
    fireEvent.click(loginBtn);

    // login failure, correct error msg displayed
    await waitFor(() => {
        expect(screen.getByText('Username Required')).toBeInTheDocument();
    });

});







