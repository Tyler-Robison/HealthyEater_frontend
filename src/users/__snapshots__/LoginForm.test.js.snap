// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="LoginForm"
  >
    <div
      class="row mt-3"
    >
      <div
        class="LoginForm-div col-12"
      >
        <h1>
          Enter Username/Password
        </h1>
        <form>
          <div>
            <div>
              <label
                for="username"
              >
                Username
              </label>
            </div>
            <input
              id="username"
              name="username"
              placeholder="Username"
              type="text"
              value=""
            />
          </div>
          <div>
            <div>
              <label
                for="password"
              >
                Password
              </label>
            </div>
            <input
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <button
            class="general-btn LoginForm-btn"
            data-testid="login-btn"
            type="submit"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
